### ISSUE TITLE
on_change method is not triggered in Streamlit 1.34

### ISSUE THREAD

**Issue by jordi-crespo** [victoryhb/streamlit-option-menu#70]
I have been working on streamlit `1.32.0` with streamlit-option-menu.

I use the `on_call` function to modify the cookies and it is working correctly.

When upgrading to streamlit `1.34.0`, this function seems not to be executed.

```python
option_menu(
                menu_title=None,
                options=self.menu_options,
                icons=self.menu_icons,
                menu_icon="cast",
                orientation="vertical",  # Cambio a vertical para la barra lateral
                default_index=default_index,  # Usar el índice de la opción guardada
                on_change=self.on_change_callback,
                key='navigation_menu',
                styles={
                    "container": {
                        "background-color": "#fff",
                        "padding": "0px",
                        "margin": "0px",
                        "border-radius": "0px"
                    },
                    "nav-link-selected": {"background-color": "#037efc", "color": "#fff"},
                    "nav-link": {"--hover-color": "#f0f2f6", "color": "##191c1e", "font-family": "'Source Sans Pro', sans-serif"},
                },
            )
            
     def handle_navigation_change(self, navigation_value):
        navigation = st.session_state.get(navigation_value)
        if navigation in self.pages or navigation == "Logout":
            self.cookie_manager.cookies["last_visited_page"] = navigation
```

**Comment by fgdvir** [comment]
I've originally added the callback from [this thread](https://github.com/streamlit/streamlit/issues/3977). It is a workaround, so it is possible that something broke in the new streamlit.
need to investigate it, but it is possible that the solution will be to remove this feature if a new workaround won't be found

**Comment by Socvest** [comment]
I posted this [here](https://discuss.streamlit.io/t/callbacks-on-custom-components-dont-seem-to-be-working-in-latest-streamlit-update/68720/2) and [here](https://discuss.streamlit.io/t/am-i-right-in-assuming-callbacks-do-not-work-for-custom-components-in-st-experimental-dialog-callbacks-broken-in-custom-components/68657) too as I noticed this

**Comment by raethlein** [comment]
The internal package structure has changed due to a refactoring; if you change the import in the `streamlit_callback.py` to `from streamlit.components.v1 import custom_component as _components`, it should work again.

That being written, this is not an official API we are going to guarantee from Streamlit-side 😉 However, we will look into this more closely and try to come up with a clean API for the future so that this patch is not needed anymore. In the meantime, I hope this^ unblocks you (for now).

**Comment by nikatlas** [comment]
I published a temporary patched version here if someone is interested from #71:

https://pypi.org/project/streamlit-option-menu-patched/0.3.12/

**Comment by raethlein** [comment]
We have also merged a PR (https://github.com/streamlit/streamlit/pull/8633) to allow passing `on_change` natively to the call of a custom component. This means, for future versions (>= 1.36) of Streamlit, the workaround patch should not be necessary anymore.

**Comment by nikatlas** [comment]
> We have also merged a PR ([streamlit/streamlit#8633](https://github.com/streamlit/streamlit/pull/8633)) to allow passing `on_change` natively to the call of a custom component. This means, for future versions (>= 1.36) of Streamlit, the workaround patch should not be necessary anymore.

Yeah I noticed that but that requires some refactoring on the streamlit-option-menu as far as I understood from your PR. Is this the case? @raethlein

**Comment by raethlein** [comment]
Yeah, you should be able to remove [these lines](https://github.com/victoryhb/streamlit-option-menu/blob/master/streamlit_option_menu/__init__.py#L59-L60) and the `register_callback` script in general and then pass the `on_change` callback to the `_component_func` call [here](https://github.com/victoryhb/streamlit-option-menu/blob/master/streamlit_option_menu/__init__.py#L65).
Since right now streamlit-option-menu documents to accept a `key` argument in the passed callback, I think the full snippet would look something like this:

```python
if on_change is not None:
    if key is None:
        st.error("You must pass a key if you want to use the on_change callback for the option menu")
    else:
        _on_change = lambda: on_change(key)

component_value = _component_func(options=options, 
            key=key, defaultIndex=default_index, icons=icons, menuTitle=menu_title, 
            menuIcon=menu_icon, default=options[default_index], 
            orientation=orientation, styles=styles, manualSelect=manual_select, on_change=_on_change)
```

**Comment by victoryhb** [comment]
Hi @raethlein @nikatlas, sorry for the late reply. I have just merged the PR by @nikatlas (thanks!) and pushed a new PyPi version. I haven't followed the recent API changes too closely. Are there any further changes that we need to make for the upcoming Streamlit releases?

**Comment by nikatlas** [comment]
> Yeah, you should be able to remove [these lines](https://github.com/victoryhb/streamlit-option-menu/blob/master/streamlit_option_menu/__init__.py#L59-L60) and the `register_callback` script in general and then pass the `on_change` callback to the `_component_func` call [here](https://github.com/victoryhb/streamlit-option-menu/blob/master/streamlit_option_menu/__init__.py#L65). Since right now streamlit-option-menu documents to accept a `key` argument in the passed callback, I think the full snippet would look something like this:
> 
> ```python
> if on_change is not None:
>     if key is None:
>         st.error("You must pass a key if you want to use the on_change callback for the option menu")
>     else:
>         _on_change = lambda: on_change(key)
> 
> component_value = _component_func(options=options, 
>             key=key, defaultIndex=default_index, icons=icons, menuTitle=menu_title, 
>             menuIcon=menu_icon, default=options[default_index], 
>             orientation=orientation, styles=styles, manualSelect=manual_select, on_change=_on_change)
> ```

@victoryhb I think the changes mentioned here are the proper solution.

**Comment by victoryhb** [comment]
I wonder if you have time to do another PR for that? @nikatlas

**Comment by nikatlas** [comment]
Cannot promise anything, not much time thats why I went with the patch. Not familiar with streamlit yet, so it will not come out of the box for me. 
But as far as I undersood the whole change is the snippet from @raethlein though it should be tested.
If my workload comes by streamlit again I will give it a try.  ⛵

**Comment by victoryhb** [comment]
Seems not much change is needed. I will do it myself. Thanks.
